infrastructure/oss/scanner.go

```diff
@@ -134,33 +144,61 @@ func (oss *Scanner) Scan(ctx context.Context, path string, _ string) (issues []s
 		workDir = filepath.Dir(path)
 	}
 
+	oss.mutex.Lock()
+	i := scanCount
+	previousScan, wasFound := oss.runningScans[workDir]
+	if wasFound && !previousScan.IsDone() { // If there's already a scan for the current workdir, we want to cancel it and restart it
+		previousScan.CancelScan()
+	}
+	newScan := scans.NewScanProgress()
+	go newScan.Listen(cancel, i)
+	scanCount++
+	oss.runningScans[workDir] = newScan
+	oss.mutex.Unlock()
+
 	cmd := oss.cli.ExpandParametersFromConfig([]string{config.CurrentConfig().CliSettings().Path(), "test", workDir, "--json"})
 	res, err := oss.cli.Execute(ctx, cmd, workDir)
+	noCancellation := ctx.Err() == nil
 	if err != nil {
-		if oss.handleError(err, res, cmd) {
+		if noCancellation {
+			if oss.handleError(err, res, cmd) {
+				return
+			}
+		} else { // If scan was cancelled, return empty results
 			return
 		}
 	}
 
-	issues = oss.unmarshallAndRetrieveAnalysis(res, uri.PathToUri(workDir))
+	issues = oss.unmarshallAndRetrieveAnalysis(ctx, res, uri.PathToUri(workDir))
+
+	oss.mutex.Lock()
+	log.Debug().Msgf("Scan %v is done", i)
+	newScan.SetDone()
+	oss.mutex.Unlock()
+
 	return issues
 }
 
 func (oss *Scanner) isSupported(documentURI sglsp.DocumentURI) bool {
 	return uri.IsDirectory(documentURI) || supportedFiles[filepath.Base(uri.PathFromUri(documentURI))]
 }
 
-func (oss *Scanner) unmarshallAndRetrieveAnalysis(res []byte, documentURI sglsp.DocumentURI) (issues []snyk.Issue) {
-	scanResults, done, err := oss.unmarshallOssJson(res)
-	if err != nil {
-		oss.errorReporter.CaptureError(err)
+func (oss *Scanner) unmarshallAndRetrieveAnalysis(ctx context.Context, res []byte, documentURI sglsp.DocumentURI) (issues []snyk.Issue) {
+	if ctx.Err() != nil {
+		return nil
 	}
 
-	if done {
-		return
+	scanResults, err := oss.unmarshallOssJson(res)
+	if err != nil {
+		oss.errorReporter.CaptureError(err)
+		return nil
 	}
 
 	for _, scanResult := range scanResults {
+		if ctx.Err() != nil {
```

Comment:
Same here: cancelling here may be of very limited value, as the conversion is done in ms.